if (!require(numbat)){
    install.packages('numbat', dependencies = TRUE,repos="https://cloud.r-project.org/")
    library(numbat)
}
library(Seurat)


features_dir <- snakemake@input[["filtered_mtx"]]
alleles_file <- paste(snakemake@input[["preproc_dir"]],"/",snakemake@params[["s_id"]],"_allele_counts.tsv.gz",sep="")
message(features_dir)
count_mat <- Read10X(data.dir=features_dir)

alleles_df <- read.csv(alleles_file,sep="\t")

# run
out = run_numbat(
    count_mat, # gene x cell integer UMI count matrix 
    ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
    alleles_df, # allele dataframe generated by pileup_and_phase script
    genome = snakemake@config[["genome_version"]],
    t = 1e-5,
    ncores = snakemake@threads,
    plot = TRUE,
    out_dir = snakemake@output[[1]],
    call_clonal_loh = snakemake@config[["numbat_loh"]],
    min_LLR = 5
)